extensions/libevent.c

```diff
@@ -0,0 +1,318 @@
+#define FUNC_NAME "primitive-add-event"
+{
+  int c_fd;
+  short c_ev;
+  struct event *event;
+  struct wait_data *data;
+  struct libevt_data *libevt;
+
+  c_fd = scm_to_int (fd);
+  c_ev = scm_to_short (ev);
+
+  libevt =
+    (struct libevt_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (0)));
+  data =
+    (struct wait_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (1)));
+
+  event = event_new (libevt->base, c_fd, c_ev, cb_func, data);
+  event_add (event, NULL);
+
+  return scm_from_pointer (event, NULL);
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_remove_event (SCM lst, SCM scm_event)
+#define FUNC_NAME "primitive-remove-event"
+{
+  struct event *event = scm_to_pointer(scm_event);
+
+  event_del (event);
+
+  return SCM_UNSPECIFIED;
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_resize (SCM lst, SCM eventsv)
+#define FUNC_NAME "primitive-resize"
+{
+  struct wait_data *data;
+
+  data =
+    (struct wait_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (1)));
+
+  data->events = (struct event_data *) SCM_BYTEVECTOR_CONTENTS (eventsv);
+  data->maxevents = SCM_BYTEVECTOR_LENGTH (eventsv) / sizeof (struct event_data);
+
+  return SCM_UNSPECIFIED;
+}
+#undef FUNC_NAME
+
+static uint64_t time_units_per_microsec;
+
+static void*
+run_event_loop (void *p)
+{
+  int microsec = 0;
+  struct timeval tv;
+
+  struct loop_data *data = p;
+
+  if (data->timeout < 0)
+    microsec = -1;
+  else if (data->timeout >= 0)
+    {
+      microsec = data->timeout / time_units_per_microsec;
+      tv.tv_sec = 0;
+      tv.tv_usec = microsec;
+    }
+
+  if (microsec >= 0)
+    event_base_loopexit (data->base, &tv);
+
+  event_base_loop (data->base, EVLOOP_ONCE);
+
+  return NULL;
+}
+
+static SCM
+scm_primitive_event_loop (SCM lst, SCM wakefd, SCM wokefd, SCM timeout)
+#define FUNC_NAME "primitive-event-loop"
+{
+  int c_wakefd;
+  int c_wokefd;
+  int result = 0;
+  int64_t c_timeout;
+
+  c_wakefd = scm_to_int (wakefd);
+  c_wokefd = scm_to_int (wokefd);
+  c_timeout = scm_to_int64 (timeout);
+
+  struct libevt_data *libevt =
+    (struct libevt_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (0)));
+  struct wait_data *data =
+    (struct wait_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (1)));
+
+  if (data == NULL)
+    scm_wrong_type_arg_msg ("event-loop", 1,
+                            scm_list_ref (lst, scm_from_int (1)),
+                            "event wait_data is null");
+
+  if (c_timeout != 0 && scm_c_prepare_to_wait_on_fd (c_wakefd))
+    data->rv = 0;
+  else
+    {
+      struct loop_data loop_data = { libevt->base, c_timeout };
+
+      if (c_timeout != 0)
+        scm_without_guile (run_event_loop, &loop_data);
+      else
+        run_event_loop (&loop_data);
+
+      if (c_timeout != 0)
+        scm_c_wait_finished ();
+
+      for (int i = 0; i < data->rv; i++)
+        {
+          if (data->events[i].fd == c_wokefd)
+            {
+              char zeroes[32];
```

Comment:
Is  the size of this significant (aside from being non-zero)? If so, or not, a comment would be appropriate.

Reply:
Not significant at all. Added a comment there and also these comment above:

```
          /* Sometimes we want to wake up the loop event and we do so by writing
           * into a write pipe (see scm_primitive_event_wake). Those writes end
           * up in the read pipe and since those are not real events we just
           * want to ignore them and drain them as we do below.
           */
```